src/Client.cpp

```diff
@@ -958,6 +989,149 @@ bool CClient::OnActionMessage(CActionMessage& Message) {
     return true;
 }
 
+void CClient::OnAuthenticateMessage(CAuthenticateMessage& Message) {
+    const auto uiMaxSaslMsgLength = 400u;
+    auto bAuthenticationSuccess = false;
+    auto sMessage = Message.GetText();
+    const auto sBufferSize = sMessage.length();
+    SCString ssMechanisms;
+
+    auto SaslReset = [this]() {
+        m_sSaslMechanism = "";
+        m_sSaslBuffer = "";
+        m_bSaslMultipart = false;
+    };
+
+    auto SaslChallenge = [this](CString sChallenge) {
+        sChallenge.Base64Encode();
+        auto sChallengeSize = sChallenge.length();
+
+        if (sChallengeSize > uiMaxSaslMsgLength) {
+            for (auto i = 0u; i < sChallengeSize; i += uiMaxSaslMsgLength) {
+                CString sMsgPart = sChallenge.substr(i, uiMaxSaslMsgLength);
+                PutClient("AUTHENTICATE " + sMsgPart);
+            }
+        } else {
+            PutClient("AUTHENTICATE " + sChallenge);
+        }
+    };
+
+    if (!m_bSasl) return;
+	
+	if (m_bSaslAuthenticated || IsAttached()) {
+        PutClient(":irc.znc.in 907 " + GetNick() +
+                  " :You have already authenticated using SASL");
+        return;
+    }
+    
+	if (!m_bSaslAuthenticating || sMessage.Equals("*")) {
+        PutClient(":irc.znc.in 906 " + GetNick() +
+                  " :SASL authentication aborted");
+        if (!IsAttached()) {
+			m_bSaslAuthenticating = false;
+            SaslReset();
+		}
+        return;
+    }
+
+    auto sMechanisms = EnumerateSaslMechanisms(ssMechanisms);
+
+	if (sBufferSize > uiMaxSaslMsgLength) {
+        PutClient(":irc.znc.in 905 " + GetNick() + " :SASL message too long");
+        SaslReset();
+        return;
+    }
+
+	if (m_sSaslMechanism.empty()) {
+        if (ssMechanisms.find(sMessage) == ssMechanisms.end()) {
+            PutClient(":irc.znc.in 908 " + GetNick() + " " + sMechanisms +
+                      " :are available SASL mechanisms");
+            PutClient(":irc.znc.in 904 " + GetNick() +
+                      " :SASL authentication failed");
+            SaslReset();
+
+            return;
+        }
+
+        m_sSaslMechanism = sMessage;
+
+        auto bResult = false;
+        CString sChallenge;
+        GLOBALMODULECALL(OnSaslServerChallenge(m_sSaslMechanism, sChallenge),
+                         &bResult);
+        if (bResult) {
+            SaslChallenge(sChallenge);
+        } else {
+            PutClient("AUTHENTICATE +");
+        }
+		return;
+	}
+
+	if (sBufferSize == uiMaxSaslMsgLength) {
+		m_bSaslMultipart = true;
+		m_sSaslBuffer.append(sMessage);
```

Comment:
Need to prevent the buffer growing too much by a sequence of multiple 400-bytes messages.